src/net_processing.cpp

```diff
@@ -2392,6 +2392,20 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                 }
                 pindexWalk = pindexWalk->pprev;
             }
+            // Special case for second cmpctblock request of tip
+            // Must match all conditions to add to vToFetch above (except mmapBlocksInFlight.count)
+            // and conditions to do CMPCT_BLOCK announcement below.
+            if (vToFetch.size() == 0 && pindexLast->pprev == chainActive.Tip() &&
+                mmapBlocksInFlight.size() == mmapBlocksInFlight.count(pindexLast->GetBlockHash()) &&
+                mmapBlocksInFlight.count(pindexLast->GetBlockHash()) < MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK &&
+                !(pindexLast->nStatus & BLOCK_HAVE_DATA) &&
+                (!IsWitnessEnabled(pindexLast->pprev, chainparams.GetConsensus()) || State(pfrom->GetId())->fHaveWitness) &&
```

Comment:
In the strange case that a peer does not set the fHaveWItness service bit, but does announce compact blocks v2, I believe this line would result in a full block request. More generally, because the two if statements always have to be in sync to avoid this, I really prefer we pull the actual request logic into this if statement.